File: ios/Classes/LocationPlusPlugin.swift
import Flutter
import UIKit
import CoreLocation

public class LocationPlusPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, LocationPlus, CLLocationManagerDelegate {
    
    private var locationManager: CLLocationManager?
    private var eventSink: FlutterEventSink?
    private var singleLocationManagers: [SingleLocationManager] = []
    
    init(channel: FlutterEventChannel) {
        super.init()
        channel.setStreamHandler(self)
    }
    
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterEventChannel(name: "haijunwei/location_plus_event", binaryMessenger: registrar.messenger())
        let instance = LocationPlusPlugin(channel: channel)
        registrar.publish(instance)
        LocationPlusSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    func startUpdatingLocation() throws {
        locationManager = CLLocationManager()
        locationManager?.delegate = self
        locationManager?.startUpdatingLocation()
    }
    
    func stopUpdatingLocation() throws {
        locationManager?.stopUpdatingLocation()
        locationManager = nil
    }
    
    func requestSingleLocation(completion: @escaping (Result<Location, Error>) -> Void) {
        let manager = SingleLocationManager(callback: completion, onDone: { [weak self] m in
            self?.singleLocationManagers.removeAll(where: { $0 === m })
        })
        manager.startUpdatingLocation()
        singleLocationManagers.append(manager);
    }
    
    
    // - MARK: CLLocationManagerDelegate
    
    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let l = locations.first else { return }
        let geocoder = CLGeocoder()
        geocoder.reverseGeocodeLocation(l, completionHandler: { [weak self] placemarks, error in
            guard let `self` = self else { return }
            if error != nil { return }
            if let e = placemarks?.first {
                let country = e.country ?? ""
                let province = e.administrativeArea ?? e.locality ?? ""
                let city = e.locality ?? ""
                let direction = e.subLocality ?? ""
                let r =  Location(latitude: l.coordinate.latitude, longitude: l.coordinate.longitude, country: country, province: province, city: city, direction: direction)
                eventSink?(r.toList())
            }
        })
    }
}


class SingleLocationManager: NSObject, CLLocationManagerDelegate {
    private let callback: (Result<Location, Error>) -> Void
    private let onDone: (SingleLocationManager) -> Void
    private let locationManager: CLLocationManager
    
    init(callback: @escaping (Result<Location, Error>) -> Void, onDone: @escaping (SingleLocationManager) -> Void) {
        self.callback = callback
        self.onDone = onDone
        locationManager = CLLocationManager()
        super.init()
        locationManager.delegate = self
    }
    
    func startUpdatingLocation() {
        locationManager.startUpdatingLocation()
    }
    
    // - MARK: CLLocationManagerDelegate
    
    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let l = locations.first else { return }
        locationManager.stopUpdatingLocation()
        let geocoder = CLGeocoder()
        geocoder.reverseGeocodeLocation(l, completionHandler: { [weak self] placemarks, error in
            guard let `self` = self else { return }
            if let e = error {
                callback(.failure(e))
                return
            }
            if let e = placemarks?.first {
                let country = e.country ?? ""
                let province = e.administrativeArea ?? e.locality ?? ""
                let city = e.locality ?? ""
                let direction = e.subLocality ?? ""
                let r = Location(latitude: l.coordinate.latitude, longitude: l.coordinate.longitude, country: country, province: province, city: city, direction: direction)
                callback(.success(r))
            } else {
                callback(.failure(LocationError.custom(msg: "反地理编码出错")))
            }
            onDone(self)
        })
    }
}

enum LocationError: LocalizedError {
    case custom(msg: String)
}

File: android/src/main/kotlin/com/haijunwei/location_plus/LocationPlusPlugin.kt
package com.haijunwei.location_plus

import android.content.Context
import com.baidu.location.BDAbstractLocationListener
import com.baidu.location.BDLocation
import com.baidu.location.LocationClient
import com.baidu.location.LocationClientOption

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel

/** LocationPlusPlugin */
class LocationPlusPlugin: FlutterPlugin, EventChannel.StreamHandler, LocationPlus {
  private var eventSink: EventChannel.EventSink? = null
  private var applicationContext: Context? = null
  private var locationClient : LocationClient? = null
  private var channel: EventChannel? = null
  private var singleLocationManagers: MutableList<SingleLocationManager> = mutableListOf<SingleLocationManager>()

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    LocationClient.setAgreePrivacy(true)
    applicationContext = flutterPluginBinding.applicationContext
    channel = EventChannel(flutterPluginBinding.binaryMessenger, "haijunwei/location_plus_event")
    channel?.setStreamHandler(this)
    LocationPlus.setUp(flutterPluginBinding.binaryMessenger, this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel?.setStreamHandler(null)
    locationClient?.stop()
    locationClient = null
  }

  override fun startUpdatingLocation() {
    locationClient = LocationClient(applicationContext)
    locationClient?.registerLocationListener(object: BDAbstractLocationListener() {
      override fun onReceiveLocation(location : BDLocation?) {
        if (location != null) {
          val l = Location(
                  location.latitude,
                  location.longitude,
                  location.country,
                  location.province,
                  location.city,
                  location.district)
          eventSink?.success(l.toList())
        }
      }
    })
    locationClient?.start()
  }

  override fun stopUpdatingLocation() {
    locationClient?.stop()
    locationClient = null
  }

  override fun requestSingleLocation(callback: (kotlin.Result<Location>) -> Unit) {
    if (applicationContext == null) { return }
    val manager = SingleLocationManager(applicationContext!!, callback) {
      singleLocationManagers.remove(it)
    }
    singleLocationManagers.add(manager)
    manager.startUpdatingLocation()
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }
}

class SingleLocationManager(
        private var context: Context,
        var callback: (Result<Location>) -> Unit,
        var onDone: (SingleLocationManager) -> Boolean) {

  private var locationClient: LocationClient = LocationClient(context)

  init {
    var option = LocationClientOption()
    option.setIsNeedAddress(true)
    locationClient.locOption = option
    locationClient.registerLocationListener(object: BDAbstractLocationListener() {
      override fun onReceiveLocation(location : BDLocation?) {
        completion(location)
      }
    });

  }

  fun startUpdatingLocation() {
    locationClient.start()
  }

  private fun completion(location : BDLocation?) {
    if (location != null) {
      val l = Location(
              location.latitude,
              location.longitude,
              location.country,
              location.province,
              location.city,
              location.district)
      callback(Result.success(l))
    } else {
      callback(Result.failure(Exception("定位出错")))
    }
    locationClient.stop()
    onDone(this)
  }
}